src/coreclr/jit/simdashwintrinsic.cpp

```diff
@@ -880,56 +880,154 @@ GenTree* Compiler::impSimdAsHWIntrinsicSpecial(NamedIntrinsic       intrinsic,
-                        op2 = gtNewSimdAsHWIntrinsicNode(retType, op2, op1, NI_SSE2_Multiply, TYP_ULONG, simdSize);
+                        case TYP_FLOAT:
+                        {
+                            hwIntrinsic = NI_SSE_Multiply;
+                            break;
+                        }
 
-                        // op2 = Sse2.Shuffle(op2, (0, 0, 2, 0))
-                        op2 = gtNewSimdAsHWIntrinsicNode(retType, op2, gtNewIconNode(SHUFFLE_XXZX, TYP_INT),
-                                                         NI_SSE2_Shuffle, baseType, simdSize);
+                        case TYP_DOUBLE:
+                        {
+                            hwIntrinsic = NI_SSE2_Multiply;
+                            break;
+                        }
+
+                        default:
+                        {
+                            unreached();
+                        }
+                    }
 
-                        // op1 = Sse2.Multiply(op1Dup.AsUInt64(), op2Dup.AsUInt64()).AsInt32()
-                        op1 =
-                            gtNewSimdAsHWIntrinsicNode(retType, op1Dup, op2Dup, NI_SSE2_Multiply, TYP_ULONG, simdSize);
+                    assert(hwIntrinsic != NI_Illegal);
+                    return gtNewSimdAsHWIntrinsicNode(retType, op1, op2, hwIntrinsic, baseType, simdSize);
+                }
 
-                        // op1 = Sse2.Shuffle(op1, (0, 0, 2, 0))
-                        op1 = gtNewSimdAsHWIntrinsicNode(retType, op1, gtNewIconNode(SHUFFLE_XXZX, TYP_INT),
-                                                         NI_SSE2_Shuffle, baseType, simdSize);
+                case NI_VectorT256_op_Multiply:
+                {
+                    NamedIntrinsic hwIntrinsic = NI_Illegal;
+                    GenTree**      broadcastOp = nullptr;
 
-                        // result = Sse2.UnpackLow(op1, op2)
-                        hwIntrinsic = NI_SSE2_UnpackLow;
+                    if (varTypeIsArithmetic(op1->TypeGet()))
+                    {
+                        broadcastOp = &op1;
                     }
-                    assert(hwIntrinsic != NI_Illegal);
+                    else if (varTypeIsArithmetic(op2->TypeGet()))
+                    {
+                        broadcastOp = &op2;
+                    }
+
+                    if (broadcastOp != nullptr)
+                    {
+                        *broadcastOp = gtNewSimdCreateBroadcastNode(simdType, *broadcastOp, baseType, simdSize,
+                                                                    /* isSimdAsHWIntrinsic */ true);
+                    }
+
+                    switch (baseType)
+                    {
+                        case TYP_SHORT:
+                        case TYP_USHORT:
+                        case TYP_INT:
+                        case TYP_UINT:
+                        {
+                            hwIntrinsic = NI_AVX2_MultiplyLow;
```

Comment:
Shouildn't this be guarded with `compOpportunisticallyDependsOn(InstructionSet_AVX2)`?

Reply:
No, its already asserted at the top of the method (https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/simdashwintrinsic.cpp#L392) as well as properly guarded further up in the stack.

We should never encounter `SimdAsHWIntrinsicClassId::VectorT256` if AVX2 is not supported.

Reply:
Why at least AVX2 is required for `VectorT256`? Shouldn't AVX be sufficient?

Reply:
`Vector<T>` supports all 10 primitive types and can only properly accelerate the integer types with AVX2 so it was decided it is only 32-bytes if AVX2 is supported and is 16-bytes otherwise, for the best overall perf/experience.